### DISCUSSION TITLE
What should I protect with AppArmor?

### DISCUSSION THREAD

**Discussion by amano-kenji** [netblue30/firejail#5434] [Q&A]
On my system, firejail is already used widely for user applications. However, for system applications, AppArmor is better.

It's just not feasible to protect every application with AppArmor.

Thus, I have to choose important targets for AppArmor protection. Can anyone give me most important protection targets?

**Comment by ghost** [comment]
Using the same guiding principles for user/system applications makes sense IMO. Trying to protect `internet-facing` or `networking` apps would be on top of my list. I recently stumbled onto [apparmor.d](https://github.com/roddhjav/apparmor.d) which offers an amazingly large collection of AA profiles.

And there's also `systemd sandboxing` for services that keeps getting tighter. But that might differ depending on your distro of choice etcetera. Beware of the potential clashes and complexities of using all these different sandboxing techniques concurrently. That in itself is not an easy job.

**Reply by curiosityseeker** [reply]
> I recently stumbled onto [apparmor.d](https://github.com/roddhjav/apparmor.d) which offers an amazingly large collection of AA profiles.

I detected that project some weeks ago, and the number of profiles is admirable. However, the project structure unfortunately breaks the AppArmor userspace tools, see https://github.com/roddhjav/apparmor.d/issues/76 and https://github.com/roddhjav/apparmor.d/issues/87 .

If you found a practicable way to overcome these problems I would be very interested ;-)

**Reply by ghost** [reply]
@curiosityseeker Yeah that's unfortunate, and i hope this gets sorted out in the future. For a while now I've been using systemd's sandboxing options for system-wide stuff instead of going full AppArmor. Mostly because there's quite some [overlap](https://github.com/netblue30/firejail/wiki/Comparison-of-firejail-and-systemd's-hardening-options) with firejail's options and I've already invested in that. But I keep eyes on the apparmor.d nonetheless and will happily share anything on that if/when improves.

**Reply by curiosityseeker** [reply]
>  For a while now I've been using systemd's sandboxing options for system-wide stuff 

Yes, I've been using systemd sandboxing as well! You might be interested in https://github.com/krathalan/systemd-sandboxing

**Reply by amano-kenji** [reply]
I don't like systemd because it claimed to be an init system, but has been trying to consume every important linux userland infrastructure.

It's an attempt at turning linux into something like microsoft windows or Mac OS. If every linux distro essentially becomes systemd, you won't really have a choice. If we don't have alternatives to systemd, redhat can try to turn systemd into a spyware or do something bad to ruin linux. MacOS reports user activities to apple, and systemd can pull off something similar if people don't have a choice.

Redhat is an unofficial government branch funded by government. What does government want? Control. They want to surveil and control everything including softwares.

I keep control agendas out of my computer. Using systemd is convenient in the same way that using credit card is convenient. Freedom is infinitely more important than convenience.

**Reply by rusty-snake** [reply]
> systemd can pull off something similar if people don't have a choice

If there is the need, people can always decide to fork systemd.

All you say is about systemd possibly becoming bad in the future. It's free software unlike MacOS, it can be forked if devs start to do bad things.

> choice

Having dev resources split over say 3 software projects doing the same is just waste of resources.

> It's an attempt at turning linux into something like microsoft windows or Mac OS.

<sup>*overstated:*</sup> Into an operating system that is used by more than a handful of computer nerds? Sounds good.

> Redhat

Also develops your Kernel.

---

What's the issue with systemd today?

**Reply by amano-kenji** [reply]
https://lists.freedesktop.org/archives/systemd-devel/2010-September/000391.html

Lennart's stance is basically redhat's stance. Lennart wrote systemd is going to eliminate other choices on all linux distributions. It's their intention.

They can try to make forks untenable by concentrating everything in systemd and developing everything really fast in systemd.

Look at chromium browser. Because it is so big and is developed so quickly, people can't fork chromium. The cost of forking chromium is extremely high for individuals and small teams. You think we can just fork chromium if alternative web browsers have been dead for a long time? Pale moon is having difficulty with staying afloat due to lack of developers.

If systemd actually becomes gigantic and eliminates competition, forking systemd can become practically impossible.

**Reply by amano-kenji** [reply]
> Having dev resources split over say 3 software projects doing the same is just waste of resources.

Waste is still better than slavery. Once people recognize slavery and agree that slavery must be abolished, then maybe it's okay to not have waste. For now, tyranny is encroaching so fast that we need any mitigations that we can get.

Slavery makes everything you have basically meaningless. For example, if your house and your car can just be taken away because you said something people in power don't like, you don't really have a house or a car. Imagine that systemd becomes a closed-source spyware that ties your computer to government ID and rats you out to people in power who will lower your social credit score if you criticize them on the internet. In the absence of alternatives, it's possible for redhat to pull it off with the help of the ruling class.

People think evil can't win. But, it can win. It had won multiple times in the 20th century. Just open a history textbook.

If slavery takes out free softwares, then you will miss the good old days of wasting resources on multiple different alternatives.

We should unite under freedom, but we must afford some waste for exiting tools of control and slavery.

**Reply by amano-kenji** [reply]
> it can be forked if devs start to do bad things.

If systemd becomes gigantic and extinguishes alternatives, developers may not want to fork systemd without being paid for maintenance.

Still, if developers unite under the understanding that slavery is bad, they can still write alternatives without being paid for maintenance.

Competing with gigantic softwares maintained by tax money when there is no alternative is going to be a real pain in the ass.

We better not lose alternatives in the first place. After we lose them, it will be really difficult to get them back. That's why I use OpenRC on linux distributions. I also use dinit for my use cases. If there is no alternative, there is no reason for systemd to be open-source. Closed-source systemd can still be maintained by tax money given to redhat.

In summary, software should stay free from centralized hierarchical control. After systemd takes out alternatives, it could become a closed-source spyware. Even open-source isn't very useful if it is under a tight grip of hierarchical control.

Stay away from (corporate) hierarchy.

**Comment by rusty-snake** [comment]
As always: "It depends on your thread model". Generally you look at the attack surface of the programs which will likely give programs handling remote/network data a higher priority. For everything else see @glitsj16 answer.

Also keep in mind that when sandboxing system services it is easy to make use system unusable and mistakes like "system does not start" are harder to fix that "music player does not start".

**Comment by CoRoe** [comment]
AppArmor allows you to impose restrictions that you cannot formulate using Firejail: Create an AppArmor profile that is applied to **all executables in the home directory** (and any other directory where the user has write access) and that **disallows everything**. Effect: No downloaded executable can do any harm.

Firejail allows to limit the functionality of specific applications (those, for which a profile is active) whereas AppArmour allows for defining a catch-all profile.
